c/frame/key.cc

```diff
@@ -0,0 +1,138 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+//------------------------------------------------------------------------------
+#include <numeric>
+#include "frame/py_frame.h"
+#include "python/list.h"
+
+
+
+//------------------------------------------------------------------------------
+// py::Frame API
+//------------------------------------------------------------------------------
+
+py::oobj py::Frame::get_key() const {
+  py::otuple key(dt->get_nkeys());
+  py::otuple names = get_names().to_pytuple();
+  for (size_t i = 0; i < key.size(); ++i) {
+    key.set(i, names[i]);
+  }
+  return std::move(key);
+}
+
+
+void py::Frame::set_key(obj val) {
+  if (val.is_none()) {
+    return dt->clear_key();
+  }
+  std::vector<size_t> col_indices;
+  if (val.is_string()) {
+    size_t index = dt->xcolindex(val);
+    col_indices.push_back(index);
+  }
+  else if (val.is_list_or_tuple()) {
+    py::olist vallist = val.to_pylist();
+    for (size_t i = 0; i < vallist.size(); ++i) {
+      py::obj item = vallist[i];
+      if (vallist[i].is_string()) {
+        size_t index = dt->xcolindex(vallist[i]);
+        col_indices.push_back(index);
+      } else {
+        throw TypeError() << "Key should be a list/tuple of column names, "
+            "instead element " << i << " was a " << item.typeobj();
+      }
+    }
+  }
+  else {
+    throw TypeError() << "Key should be a column name, or a list/tuple of "
+        "column names, instead it was a " << val.typeobj();
+  }
+  _clear_types();
+  dt->set_key(col_indices);
+}
+
+
+
+//------------------------------------------------------------------------------
+// DataTable API
+//------------------------------------------------------------------------------
+
+size_t DataTable::get_nkeys() const {
+  return nkeys;
+}
+
+void DataTable::clear_key() {
+  nkeys = 0;
+}
+
+
+void DataTable::set_key(std::vector<size_t>& col_indices) {
+  if (col_indices.empty()) {
+    nkeys = 0;
+    return;
+  }
+  // Check that col_indices are unique
+  size_t K = col_indices.size();
+  for (size_t i = 0; i < K; ++i) {
+    for (size_t j = i + 1; j < K; ++j) {
+      if (col_indices[i] == col_indices[j]) {
+        throw ValueError() << "Column `" << names[col_indices[i]] << "` is "
+          "specified multiple times within the key";
+      }
+    }
+  }
+
+  // Sort the table by the keys
+  Groupby gb;
+  RowIndex ri = sortby(col_indices, &gb);
+  xassert(ri.length() == nrows);
```

Comment:
What are the situations when `ri.length() == nrows` is false?

Reply:
Well, this is an assertion, which means there shouldn't be any.

Reply:
Right, but I was like this is already checked by the `sortby` function itself?